File: P3/Napakalaki/src/napakalaki/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package napakalaki;
import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author Victor Vallecillo Morilla
 */
public class Player 
{
   public static int MAXLEVEL=10;
   private String name;
   private int level;
   private boolean dead=true;
   private boolean canISteal=true;
   private Player enemy;
   private ArrayList<Treasure> visibleTreasures=new ArrayList();
   private ArrayList<Treasure> hiddenTreasures=new ArrayList();
   private BadConsequence pendingBadConsequence=new BadConsequence("",0,0,0);
 
   public Player(String name)
   {
       this.name=name;
      
   }
   
   public String getName()
   {
       return name;
   }
   private void bringToLife()
   {
       dead=true;
   }
   public int getCombatLevel()
   {
       int total=0;
       
       for( Treasure t: hiddenTreasures)
           total=total+t.getBonus();
       
       return (total+level);
       
   }
   
   public static int getMAXLEVEL() {
        return MAXLEVEL;
   }

   public int getLevel() {
        return level;
   }

   public boolean isCanISteal() {
        return canISteal;
   }

   public Player getEnemy() {
        return enemy;
   }
   
   private void incrementLevels(int i)
   {
       level+=i;
       if(level> MAXLEVEL)
               level=10;
   }
   private void decrementLevels(int i)
   {
       level-=i;
       
       if(level <0)
           level=0;
   }
   private void setPendingBadConsequence( BadConsequence b)
   {
       pendingBadConsequence=b;
   }  
   
   public void applyPrize(Monster m)   // MODIFICADO
   {
       int nLevels=m.getLevelsGained();
       incrementLevels(nLevels);
       int nTreasures=m.getTreasuresGained();
       
       if(nTreasures>0)
       { 
           CardDealer dealer=CardDealer.getInstance();
           
           for(int i=0;i<nTreasures;i++)
           {
               Treasure treasure=dealer.nextTreasure();
               hiddenTreasures.add(treasure);
           }
       }
   }
   public void applyBadConsequence(Monster m) // MODIFICADO
   {
       BadConsequence badConsequence=m.getBadConsequence();
       int nLevels=badConsequence.getLevels();
       
       decrementLevels(nLevels);
       
       
       BadConsequence pendingBad=badConsequence.adjustToFitTreasureLists(this.visibleTreasures, this.hiddenTreasures);
       
       setPendingBadConsequence(pendingBad);
   }
   private boolean canMakeTreasureVisible(Treasure t)    // MODIFICADO
   {
       boolean resultado = true;
        if (visibleTreasures.size() < 4)
        {
            TreasureKind tipo = t.getType();
            if (tipo != TreasureKind.ONEHAND && tipo != TreasureKind.BOTHHANDS)
            {
                boolean hasOne = false; 
                for (Treasure tv : visibleTreasures)
                {
                    if (tv.getType() == tipo)
                        hasOne = true; 
                }
                return !hasOne; 
            }
            else if (tipo == TreasureKind.ONEHAND)
            {
                int i = 0; 
                boolean hasBothhands = false; 
                for (Treasure Vt : visibleTreasures)
                {
                    if(Vt.getType() == TreasureKind.ONEHAND)
                        i++;
                    else if (Vt.getType() == TreasureKind.BOTHHANDS)
                        hasBothhands = true; 
                }   
                return (i < 2&&!hasBothhands); 
            }     
            else
            {
                boolean Both = false; 
                boolean onehands = false; 
                for (Treasure Vt : visibleTreasures)
                {
                    if(Vt.getType() == TreasureKind.ONEHAND)
                        onehands = true;
                    else if (Vt.getType() == TreasureKind.BOTHHANDS)
                        Both = true; 
                }
                return (!onehands && !Both );
            }
        }
        return resultado;
       
   }
   private int howManyVisibleTreasures(TreasureKind tKind)
   {
       int contador=0;
       
       for(Treasure t: visibleTreasures)
            if(t.getType()==tKind)
               contador++;
       
       return contador;       
   }
   private void dieIfNoTreasures()
   {
       if(hiddenTreasures.isEmpty() && visibleTreasures.isEmpty())
           dead=true;
   }
   public boolean isDead()
   {
      return dead; 
   } 
   
   public ArrayList<Treasure> getVisibleTreasures() {
        return visibleTreasures;
   }

   public ArrayList<Treasure> getHiddenTreasures() {
        return hiddenTreasures;
   }

   public BadConsequence getPendingBadConsequence() {
        return pendingBadConsequence;
   }

   public CombatResult combat(Monster m)
   {
       CombatResult combat;
       int myLevel =getCombatLevel();
        int monsterLevel=m.getCombatLevel();
        
        if(myLevel > monsterLevel)
        {
           applyPrize(m);   
           
           if(getLevels()>=MAXLEVEL)
              combat=CombatResult.WINGAME;
           else
               combat=CombatResult.WIN;

        }
        else
        {
            applyBadConsequence(m);
            combat=CombatResult.LOSE;
        }

        return combat;
   }
   public void makeTreasureVisible(Treasure t)     
   {
       boolean CanI=canMakeTreasureVisible(t);
       if(CanI)
       {
           visibleTreasures.add(t);
           hiddenTreasures.remove(t);
       }
          
   }
   public void discardVisibleTreasure(Treasure t)    
   {
       visibleTreasures.remove(t);
       
       if ((this.pendingBadConsequence != null) && (!this.pendingBadConsequence.isEmpty()))
            pendingBadConsequence.substractVisibleTreasure(t);
       
       this.dieIfNoTreasures();
   }
   public void discardHiddenTreasure(Treasure t)   
   {
         hiddenTreasures.remove(t);
         
       if (this.pendingBadConsequence != null && !this.pendingBadConsequence.isEmpty())
             pendingBadConsequence.substractHiddenTreasure(t);
       
       CardDealer.getInstance().giveTreasureBack(t); 
       this.dieIfNoTreasures();
   }
   public boolean validState()
   {
       boolean valido=false;
       
     if(pendingBadConsequence.isEmpty() && hiddenTreasures.size()<=4)
         valido=true;
     
     return valido;
   }
   public void initTreasures()         
   {
       CardDealer dealer=CardDealer.getInstance();
       Dice dice=Dice.getInstance();
       bringToLife();
       
       Treasure treasure=dealer.nextTreasure();
       
       hiddenTreasures.add(treasure);
       int number=dice.nextNumber();
       
       if(number>1 )
       {
           treasure=dealer.nextTreasure();
           hiddenTreasures.add(treasure);
       }
       if(number==6)
        {
             treasure=dealer.nextTreasure();
           hiddenTreasures.add(treasure);
        }
   }
   public int getLevels()
   {
     return level;
   }
   public Treasure stealTreasure()
   {
       Treasure treasure=null;
      canISteal=canISteal();
      if(canISteal)
      {
          boolean canYou=enemy.canYouGiveMeATreasure();
          
          if(canYou)
          {
              treasure=enemy.giveMeATreasure();
              hiddenTreasures.add(treasure);
              haveStolen();
          }
      }
      return treasure;
   }
   public void setEnemy(Player enemy)
   {
        this.enemy=enemy;
   }
   private Treasure giveMeATreasure()   // MODIFICADO
   {
     Random r = new Random();
     int tope=hiddenTreasures.size();
     int valor= r.nextInt(tope);
     
     return hiddenTreasures.get(valor);
   }
   public boolean canISteal() 
   {
    return canISteal;
   }
   private boolean canYouGiveMeATreasure()
   {
     boolean ser_robado=false;
     if(visibleTreasures.size()>0 || hiddenTreasures.size()>0)
         ser_robado=true;
     
     return ser_robado;  
   }
   private void haveStolen()  
   {
      // if(canISteal()==false)
           canISteal=false;
   }
   public void discardAllTreasures()  // MODIFICADO
   {
       while(!visibleTreasures.isEmpty())
         discardVisibleTreasure(visibleTreasures.get(0));
       
         
       while(!hiddenTreasures.isEmpty())
        discardHiddenTreasure(hiddenTreasures.get(0));
       
       
   } 
   
   @Override
   public String toString()
   {
       return "\n Nombre: "+name+"\n Nivel: "+level ;
   }
    
}

File: P3/NapakalakiRuby/lib/player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
 
  require_relative 'combat_result'
  require_relative 'treasure_kind'
  require_relative 'dice'
  require_relative 'treasure'
  require_relative 'monster'
  
  
module NapakalakiRuby

  class Player
    attr_reader :name, :hiddenTreasures, :visibleTreasures , :level, :dead,
     :pendinBadConsequence, :enemy
    @@MAXLEVEL=10    #como es de clase se inicializa fuera
    @@MINLEVEL=0
    
    #no se de donde sale combatLevels, lo quito

    def initialize(name)
      @name=name
      @dead=true
      @level=1
      @visibleTreasures=Array.new
      @hiddenTreasures=Array.new
      @enemy
      @pendinBadConsequence
      @canIsteal=true
     end

    def bringToLife
      @dead=false;

    end

    def incrementLevels(l)
      @level=@level+l
      
      if(@level>@@MAXLEVEL)
        @level=10
      end
    end

    def decrementLevels(l)
      @level=@level-l
      
      if @level<0
        @level=0
      end
    end

    def setPendingBadConsequence(b)
      @pendinBadConsequence=b
    end
    #Modificado S2
    def  applyPrize(m)
      nLevels=m.getLevelsGained
      incrementLevels(nLevels)
      nTreasures=m.getTreasuresGained

      if nTreasures > 0 then
        dealer=CardDealer.instance
        nTreasures.each do |i|
          treasure=dealer.nextTreasure
          @hiddenTreasures << treasure
        end
      end
    end
    #Modificado S2
    def applyBadConsequence(m)
      badConsequence=m.badConsequence
      nLevels=badConsequence.levels
      decrementLevels(nLevels)
      @pendingBadConsequence=badConsequence.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
      setPendingBadConsequence(@pendinBadConsequence)
    end
    #Modificado
    def canMakeTreasureVisible(t)
      cont=0
      i=0
      correcto=false
      while i < @visibleTreasures.size
        if @visibleTreasures[i].type== t.type
          cont=cont+1
        end
        i=i+1
      end
      if cont < 1 || t.type== TreasureKind::ONEHAND && cont<2
        correcto=true
      end
      return correcto
    end

    def howManyVisibleTreasures(tKind)
      @contador=0

        @hiddenTreasures.each{ |h| 
          if h.type == tKind
            @contador=@ontador+1
          end
        }

      contador
    end

    def dieIfNoTreasures
      if(@hiddenTreasures.empty? && @visibleTreasures.empty?)
            @dead=true;
      end
    end

    def isDead
      @dead
    end

    def combat(m)
      myLevel=@level
      monsterLevel=m.combatLevel
      
      if myLevel > monsterLevel then
        applyPrize(m)
        if @level >= MAXLEVEL then
          applyPrize(m)
          combat=CombatResult::WINGAME
        else
          combat=CombatResult::WIN
        end
      else
        applyBadConsequence(m)
        combat=CombatResult::LOSE
      end
      return combat
    end

    def makeTreasureVisible(t)
      canI=canMakeTreasureVisible(t)
      if canI then
        @visibleTreasures << t #o usamos .push(t)
        @hiddenTreasures.delete(t)        
      end
    end

    def discardVisibleTreasure(treasure)
      @hiddenTreasures.delete(treasure)
      if @pendingBadConsequence != nil  && !@pendingBadConsequence.empty?
        @pendinBadConsequence.substractVisibleTreasure(treasure)

      end
      CardDealer.instance.giveTreasureBack(treasure)
      dieIfNoTreasures
    end

    def discardHiddenTreasure(treasure)
        @hiddenTreasures.delete(treasure)

      if @pendinBadConsequence != nil && !@pendinBadConsequence.empty?
        @pendinBadConsequence.substractHiddenTreasure(treasure)
      end
      CardDealer.instance.giveTreasureBack(treasure)
      dieIfNoTreasures
    end

    def validState

      valido=false;

       if @pendingBadConsequence == nil && @hiddenTreasures.lenght <= 4
           valido=true;
       end 

        valido

    end
    #Modificado S2
    def initTreasures
      dealer=CardDealer.instance
      dice=Dice.instance
      bringToLife
      treasure=dealer.nextTreasure
      @hiddenTreasures << treasure
      number=dice.nextNumber
      
      if number > 1 then
        treasure=dealer.nextTreasure
        @hiddenTreasures << treasure
      end
      
      if number == 6 then
        treasure=dealer.nextTreasure
        @hiddenTreasures << treasure
      end
    end

    def getLevels
      return @level
    end
    #Modificado S2
    def stealTreasure
      treasure=nil
      canI=canISteal
      if canI then
        canYou=@enemy.canYouGiveMeATreasure
        if canYou
          treasure=@enemy.giveTreasureBack
          @hiddenTreasures << treasure
          haveStolen
        end
      end
      return treasure
    end

    def setEnemy( enemy)
      @enemy=enemy  

    end
    #Modificado
    def giveMeATreasure
      totTesOcu=@hiddenTreasures.length
      valor=rand(totTesOcu)

      return @hiddenTreasures.at(valor)
    end

    def canISteal
     robado=false

      if @hiddenTreasures.empty? && @visibleTreasures.empty?
        robado=true
       end

       return robado
    end

    def canYouGiveMeATreasure

        ser_robado=false;
        if @visibleTreasures.size >0 || @hiddenTreasures.size >0
           ser_robado=true;
       end
       return ser_robado
    end

    def haveStolen
      @canISteal=false
    end
    #Modificado S2 Copia de colecciones
    def discardAllTreasures
      
      while !visibleTreasures.empty?
        discardVisibleTreasure(visibleTreasures.at(0))
      end
      
      while !hiddenTreasures.empty?
        discardHiddenTreasure(hiddenTreasures.at(0))
      end
      
    end
    
    def to_s
      " Nombre: #{@name} \nNivel: #{@level}"
    end
    
  end
end
